### PR TITLE
Emit copy when calling non-readonly method with ref readonly field receiver

### LINKED ISSUE #66128: Bug: Using byref fields, it's possible to mutate a ref readonly without any unsafe code
**Version Used**:

Reproduces on both:
- Sharplab.io
- VS 17.4.0 with .NET 7.0.100 SDK

**Steps to Reproduce**:

1. Create a `ref struct` with a `ref readonly` field to the struct you want to accidentally mutate
2. Instantiate an instance of this new ref struct type with a `ref readonly` or `in` parameter that you want to mutate
3. Call a mutating (not extension) method on the value through the field in the ref struct instance

Minimal repro:
https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK4wD4AEBMBGAWACgsBmAAl3IGFiBvY8pynHR5ho578gZXIBuEADaoApuQC85GGIDuACgCUAbnY9yyMQDNNYiABMA9jGEBPPoJHiASjql7dQ0WLVcNlPAE4Fz2ztV1HgBZBVgrFzttQPcNLG9fazEomO4AXyCyTwA2SgAWclDw/j8xJSDOD3IbTQdZRXDSmKDuAHpWgEkAZy7xcigu8gQACwkRiQBjIzAAByhhMWRyYzEumAByBD0RAbHhiC315AA6ADEoMWEDdf7BnW0xCYQoAUuLLUMTczgW5nauoz9LayMQGQYIQETD4ICQQZY6MQwLovSZGGYWIy6cZ8PARPoQGAGcgTETCQrKcgmIb7BDHX5ME7nS4GY6hVIeenkdoAFVGxKMBjEAHNEeRtEYliMBsTSbd+cgtE9yMiYBMJMhgnK4R9jKYLGAxCMBT9Yjx2gSiSNUIM5PMybJXkswAKoNoLNjGRcro5Foi1etBqU6abuJ7mccNco3OliBkSBQugg0ErePQgrAtlBo8wslgCmyKpyoABqYvZphx7hZHVfCx5zRs8h0ciVnMUGt6ymO5BQQWeAAM5G5Rl4SdgQopkgAfP1jsPRz2YBOYnHq/ZE8mtjY06a17oO+ZLEyruXKBQbGEYJYmlIZ8eidItE4km40kA==

**Expected Behavior**:

Compiler creates a defensive copy to the value from `r.Field` before calling `M()` on it since `M()` is not a readonly method.

**Actual Behavior**:

Compiler calls directly on the ref from `r.Field`, and thus allowing mutation of a `ref readonly`. The behaviour for calling `rM()` is correct.

### PR DESCRIPTION
**Pull request by cston**
Fixes #66128

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs [hunk 1]
```diff
@@ -4641,7 +4641,7 @@ internal enum AddressKind
                 return false;
             }
 
-            if (field.RefKind is RefKind.Ref or RefKind.RefReadOnly)
+            if (field.RefKind is RefKind.Ref)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs:4644]: >if (field.RefKind is RefKind.Ref) [](http://example.com/codeflow?start=12&length=33)

It looks like doc comment doesn't reflect the conditions accurately.

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs [hunk 2]
```diff
@@ -4664,7 +4664,9 @@ internal enum AddressKind
                 return false;
             }
 
-            if (!field.IsReadOnly)
+            Debug.Assert(field.RefKind is RefKind.None or RefKind.RefReadOnly);
+
+            if (field.RefKind == RefKind.None && !field.IsReadOnly)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs:4669]: >field.RefKind == RefKind.None [](http://example.com/codeflow?start=16&length=29)

It looks like at this point we are always supposed to return `false` for ```RefKind.RefReadOnly```. Consider doing that explicitly and reverting changes below.

File: src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs:4644]: For example, it appear a ref field has a home even when it is readonly.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 1)

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 4)

**Comment by cston** [comment]
@dotnet/roslyn-compiler, please review, thanks.

**Review by RikkiGibson — APPROVED** [review]

**Pull request merged**